processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.groupby;
+
+import org.apache.druid.guice.LazySingleton;
+import org.apache.druid.query.groupby.epinephelinae.LimitedTemporaryStorage;
+
+import javax.inject.Inject;
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Collects stats for group by queries like used merged buffer count, spilled bytes and group by resource acquisition time.
+ */
+@LazySingleton
+public class GroupByStatsProvider
+{
+  private final AtomicLong resourceAcquisitionTimeNs = new AtomicLong(0);
+  private final AtomicLong resourceAcquisitionCount = new AtomicLong(0);
+
+  private final ConcurrentLinkedQueue<LimitedTemporaryStorage> temporaryStorages;
+
+  @Inject
+  public GroupByStatsProvider()
+  {
+    this.temporaryStorages = new ConcurrentLinkedQueue<>();
+  }
+
+  public synchronized void groupByResourceAcquisitionTimeNs(long delayNs)
+  {
+    resourceAcquisitionTimeNs.addAndGet(delayNs);
+    resourceAcquisitionCount.incrementAndGet();
+  }
+
+  public synchronized long getAndResetGroupByResourceAcquisitionStats()
+  {
+    long average = resourceAcquisitionCount.get() != 0 ?
+                   (resourceAcquisitionTimeNs.get() / resourceAcquisitionCount.get()) : 0;
+
+    resourceAcquisitionTimeNs.set(0);
+    resourceAcquisitionCount.set(0);
+
+    return average;
+  }
+
+  public void registerTemporaryStorage(LimitedTemporaryStorage temporaryStorage)
```

Comment:
How often does this method need to be called per group by query, and what does this rely upon? I feel that there's an inbuilt expense with tracking the spilled bytes this way. I think it is still fine, but I wonder if a more efficient way would be as follows:

1. Have a running counter of the size occupied 
2. LimitedTemporaryStorage implements closeable. Modify wherever the LimitedTemporaryStorage is getting closed to subtract the value of the storage from the counter. There can be a set inclusion check where a hashset of the opened temporary storages being tracked by the running counter are stored, and the code confirms that the removed temporary storage was indeed being tracked by the counter. 

Note: My approach seems more convoluted, so if there isn't much performance downside of the current version, I think its fine as is.

Reply:
>there's an inbuilt expense with tracking the spilled bytes this way

What is the expense you are referring to? The temporary storage object is added to a queue and cost of adding an element should be in most cases should be O(1). 

>LimitedTemporaryStorage implements closeable. Modify wherever the LimitedTemporaryStorage is getting closed to subtract the value of the storage from the counter. There can be a set inclusion check where a hashset of the opened temporary storages being tracked by the running counter are stored, and the code confirms that the removed temporary storage was indeed being tracked by the counter.

This approach also requires keeping track of the temporary storages thus incurring similar overhead as the existing implementation. Additionally, it requires incrementing the "running counter" whenever bytes are written out to the disk.

Reply:
> What is the expense you are referring to

Iterating while calculating the stored bytes.